incentivization.md

```diff
@@ -0,0 +1,206 @@
+
+## Waku Store
+
+In this document, we focus on i13n for Waku Store.
+
+Store is a client-server protocol that currently works as follows:
+1. the client sends a `HistoryQuery` to the server;
+2. the server sends a `HistoryResponse` to the client.
+
+The response may be split into multiple parts, as specified by pagination parameters in `PagingInfo`.
+
+Let us define a relevant message as a message that has been broadcast via Relay within the time frame and matching the filter criteria that the client specified.
+The desired functionality of Store can be described as following:
+
+- the server responds quickly;
+- all relevant messages are in the response;
+- the response contains only relevant messages.
+
+# Waku Store incentivization MVP
+
+We propose to add the following aspects to the protocol:
+1. pricing:
+	1. cost calculation
+	2. price advertisement
+	3. price negotiation
+2. payment:
+	1. payment itself
+	2. proof of payment
+3. reputation
+4. results cross-checking
+
+In this document, we define the simplest viable i13n modification to the Store protocol (MVP).
+The MVP protocol has no price advertisement, no price negotiation, and no results cross-checking.
+Other elements are present in a minimal version.
+In further subsections, we list the potential direction for future work towards a fully-fledged i13n protocol.
+
+## Pricing
+
+For MVP, we assume a constant price per hour of history.
+After the client sends a `HistoryQuery` to the server:
+1. The server internally calculates the offer price and sends it to the client.
+2. If the client agrees, it pays and sends a proof of payment to the server.
+3. If the client does not agree, it sends a rejection message to the server.
+4. If the server receives a valid payment before a certain timeout, it sends the response to the client.
+5. If the server receives a rejection message, or receives no message before a timeout, the server assumes that the client has rejected the offer.
+
+### Future work
+
+- DoS protection: a client can overwhelm a server with requests and not proceed to payment. Countermeasure: ignore requests from the same client if they come too often; generalize a reputation system to servers ranking clients.
+- Cost calculation - see https://github.com/waku-org/research/issues/35
+- Price advertisement - see https://github.com/waku-org/research/issues/51
+- Price negotiation - see https://github.com/waku-org/research/issues/52
+
+## Payment
+
+If the client agrees to the price, it sends a _proof of payment_ to the server.
+For the MVP, each request is paid for with a separate blockchain transaction.
+The transaction hash (`txid`) acts as a proof of payment.
+
+Note that client gives proof of payment before it receives the response.
+Other options could be:
+1. the client pays after the fact;
+2. the client pays partly upfront and partly after the fact;
+3. an escrow (a centralized trusted third party, or a semi-trusted entity like a smart contract) ensures atomicity .
```

Comment:
Is it possible to ensure atomicity without an escrow?

Reply:
Good question! Depends on how you define escrow. But I'll add a few words along the lines of "atomic swaps", Lightning / Hopr (thanks for the link to Hopr), etc.